kiwipy/communicate.py

```diff
@@ -0,0 +1,12 @@
+__all__ = ('connect',)
+
+DEFAULT_COMM_URI = 'amqp://guest:guest@127.0.0.1/'
+
+
+def connect(uri: str = DEFAULT_COMM_URI, **kwargs):
```

Comment:
should this become a proper context manager function that automatically closes the communicator? I see in the examples you are now calling it as
```
with kiwipy.connect() as comm:
    # use comm
```
but I don't see what the benefit there is of using it in a context, or am I missing something?

Reply:
The benefit is the auto closing (and being explicit about the lifetime).  But it's not always possible to use context managers.  They're somewhat of a `python` hack for the fact that it doesn't have proper object lifetime management (via, say, RAII).  `connect()` isn't a cm for (I guess) similar reasons to why `python`'s `open()` isn't a cm.  There are just times when scope can't be used to restrict the lifetime of these kinds of objects e.g. when doing ownership transfer to a function or object that will itself do the closing (or so you hope!).  I think I may even do this in some of my code where one thing knows how to create communicators and the other just wants to consume it via dependency injection.

Reply:
that is my point though, the `Communicator` has a `close` method to clean up the resources no? Wouldn't it be the idea to implement `connect` as a contextmanager that constructs the `Communicator`, yields it, and then calls close when the caller exits the context. Something like
```
import contextlib

@contextlib.contextmanager
def connect(params):
    communicator = Communicator.connect(params)
    try:
        yield communicator
    finally:
        communicator.close()
```

as it stands, the `with kiwipy.connect()` you are using in the example does not do anything right? Just getting rid of the `with` there would not change anything at all, or am I missing something?

Reply:
Ah, I see the confusion.  So `RmqCommunicator` is a context manager which is why the context even worked.  It was my intention for all `Communicator`s to have this functionality but I hadn't promoted it to the interface.  Done now, also discovered a bug in the RMQ communicator that I fixed with this latest commit.